### ISSUE TITLE
训练自己的VOC数据集时，出现TypeError: __init__() got an unexpected keyword argument 'EvalReader'

### ISSUE THREAD

**Issue by Lxiaom123** [PaddlePaddle/PaddleDetection#1589]
![image](https://user-images.githubusercontent.com/66930694/96688056-499c8400-13b3-11eb-9cd5-d92307d964da.png)
yml文件如下

```
architecture: FasterRCNN
max_iters: 66000
use_gpu: false
snapshot_iter: 10000
log_smooth_window: 20
save_dir: output
pretrain_weights: https://paddle-imagenet-models-name.bj.bcebos.com/ResNet50_cos_pretrained.tar
metric: VOC
map_type: 11 point
weights: output/faster_rcnn_r50_fpn_1x/model_final
num_classes: 4

FasterRCNN:
  backbone: ResNet
  fpn: FPN
  rpn_head: FPNRPNHead
  roi_extractor: FPNRoIAlign
  bbox_head: BBoxHead
  bbox_assigner: BBoxAssigner

ResNet:
  norm_type: bn
  norm_decay: 0.
  depth: 50
  feature_maps: [2, 3, 4, 5]
  freeze_at: 2

FPN:
  min_level: 2
  max_level: 6
  num_chan: 256
  spatial_scale: [0.03125, 0.0625, 0.125, 0.25]

FPNRPNHead:
  anchor_generator:
    anchor_sizes: [32, 64, 128, 256, 512]
    aspect_ratios: [0.5, 1.0, 2.0]
    stride: [16.0, 16.0]
    variance: [1.0, 1.0, 1.0, 1.0]
  anchor_start_size: 32
  min_level: 2
  max_level: 6
  num_chan: 256
  rpn_target_assign:
    rpn_batch_size_per_im: 256
    rpn_fg_fraction: 0.5
    rpn_positive_overlap: 0.7
    rpn_negative_overlap: 0.3
    rpn_straddle_thresh: 0.0
  train_proposal:
    min_size: 0.0
    nms_thresh: 0.7
    pre_nms_top_n: 2000
    post_nms_top_n: 2000
  test_proposal:
    min_size: 0.0
    nms_thresh: 0.7
    pre_nms_top_n: 1000
    post_nms_top_n: 1000

FPNRoIAlign:
  canconical_level: 4
  canonical_size: 224
  min_level: 2
  max_level: 5
  box_resolution: 7
  sampling_ratio: 2

BBoxAssigner:
  batch_size_per_im: 512
  bbox_reg_weights: [0.1, 0.1, 0.2, 0.2]
  bg_thresh_lo: 0.0
  bg_thresh_hi: 0.5
  fg_fraction: 0.25
  fg_thresh: 0.5

BBoxHead:
  head: TwoFCHead
  nms:
    keep_top_k: 100
    nms_threshold: 0.5
    score_threshold: 0.05

TwoFCHead:
  mlp_dim: 1024

LearningRate:
  base_lr: 0.0025
  schedulers:
  - !PiecewiseDecay
    gamma: 0.1
    milestones: [40000, 54000]
  - !LinearWarmup
    start_factor: 0.1
    steps: 1000

OptimizerBuilder:
  optimizer:
    momentum: 0.9
    type: Momentum
  regularizer:
    factor: 0.0001
    type: L2

_READER_: 'faster_fpn_reader.yml'

TrainReader:
  inputs_def:
    fields: ['image', 'im_info', 'im_id', 'gt_bbox', 'gt_class','is_difficult']
  dataset:
    !VOCDataSet
    dataset_dir: /home/aistudio/work/CCTSDB_VOC1/VOCdevkit/VOC2012/
    anno_path: /home/aistudio/work/CCTSDB_VOC1/trainval.txt
    use_default_label: false
  #batch_size: 2

  EvalReader:
  inputs_def:
    fields: ['image', 'im_info', 'im_id', 'im_shape', 'gt_bbox', 'gt_class', 'is_difficult', 'is_crowd']
  dataset:
    !VOCDataSet
    dataset_dir: /home/aistudio/work/CCTSDB_VOC1/VOCdevkit/VOC2012/
    anno_path: /home/aistudio/work/CCTSDB_VOC1/test.txt
    use_default_label: false

TestReader:
  dataset:
    !ImageFolder
    anno_path: /home/aistudio/work/CCTSDB_VOC1/label_list.txt
    use_default_label: false
```

**Comment by willthefrog** [comment]
如果要自定义reader，需要删除这行。
```yaml
READER: 'faster_fpn_reader.yml'
```

**Comment by Lxiaom123** [comment]
faster_fpn_reader.yml我把这个文件内容改了，文件名保持不变，不行吗？

**Comment by Lxiaom123** [comment]
faster_fpn_reader.yml内容如下
TrainReader:
  inputs_def:
    fields: ['image', 'im_info', 'im_id', 'gt_bbox', 'gt_class','is_difficult']
  dataset:
    !VOCDataSet
    dataset_dir: /home/aistudio/work/CCTSDB_VOC1
    anno_path: /home/aistudio/work/CCTSDB_VOC1/trainval.txt
    use_default_label: false
  sample_transforms:
  - !DecodeImage
    to_rgb: true
  - !RandomFlipImage
    prob: 0.5
  - !NormalizeImage
    is_channel_first: false
    is_scale: true
    mean: [0.485,0.456,0.406]
    std: [0.229, 0.224,0.225]
  - !ResizeImage
    target_size: 800
    max_size: 1333
    interp: 1
    use_cv2: true
  - !Permute
    to_bgr: false
    channel_first: true
  batch_transforms:
  - !PadBatch
    pad_to_stride: 32
    use_padded_im_info: false
  batch_size: 1
  shuffle: true
  worker_num: 2
  use_process: false

EvalReader:
  # inputs_def:
  #   #fields: ['image', 'im_info', 'im_id', 'im_shape']
  #   # for voc
  #   fields: ['image', 'im_info', 'im_id',  'im_shape', 'gt_bbox', 'gt_class', 'is_difficult']
  # dataset:
  #   !COCODataSet
  #   image_dir: val2017
  #   anno_path: annotations/instances_val2017.json
  #   dataset_dir: dataset/coco
  inputs_def:
    fields: ['image', 'im_info', 'im_id',  'im_shape', 'gt_bbox', 'gt_class', 'is_difficult', 'is_crowd']
  dataset:
    !VOCDataSet
    dataset_dir: /home/aistudio/work/CCTSDB_VOC1
    anno_path: /home/aistudio/work/CCTSDB_VOC1/test.txt
    use_default_label: false
  sample_transforms:
  - !DecodeImage
    to_rgb: true
    with_mixup: false
  - !NormalizeImage
    is_channel_first: false
    is_scale: true
    mean: [0.485,0.456,0.406]
    std: [0.229, 0.224,0.225]
  - !ResizeImage
    interp: 1
    max_size: 1333
    target_size: 800
    use_cv2: true
  - !Permute
    channel_first: true
    to_bgr: false
  batch_transforms:
  - !PadBatch
    pad_to_stride: 32
    use_padded_im_info: true
  batch_size: 1
  shuffle: false
  drop_empty: false
  worker_num: 2

TestReader:
  inputs_def:
    # set image_shape if needed
    fields: ['image', 'im_info', 'im_id', 'im_shape']
  # dataset:
  #   !ImageFolder
  #   anno_path: annotations/instances_val2017.json
  dataset:
    !ImageFolder
    anno_path: /home/aistudio/work/CCTSDB_VOC1/label_list.txt
    use_default_label: false
  sample_transforms:
  - !DecodeImage
    to_rgb: true
    with_mixup: false
  - !NormalizeImage
    is_channel_first: false
    is_scale: true
    mean: [0.485,0.456,0.406]
    std: [0.229, 0.224,0.225]
  - !ResizeImage
    interp: 1
    max_size: 1333
    target_size: 800
    use_cv2: true
  - !Permute
    channel_first: true
    to_bgr: false
  batch_transforms:
  - !PadBatch
    pad_to_stride: 32
    use_padded_im_info: true
  batch_size: 1
  shuffle: false

**Comment by Lxiaom123** [comment]
faster_fpn_reader.yml内容如下
TrainReader:
  inputs_def:
    fields: ['image', 'im_info', 'im_id', 'gt_bbox', 'gt_class','is_difficult']
  dataset:
    !VOCDataSet
    dataset_dir: /home/aistudio/work/CCTSDB_VOC1
    anno_path: /home/aistudio/work/CCTSDB_VOC1/trainval.txt
    use_default_label: false
  sample_transforms:
  - !DecodeImage
    to_rgb: true
  - !RandomFlipImage
    prob: 0.5
  - !NormalizeImage
    is_channel_first: false
    is_scale: true
    mean: [0.485,0.456,0.406]
    std: [0.229, 0.224,0.225]
  - !ResizeImage
    target_size: 800
    max_size: 1333
    interp: 1
    use_cv2: true
  - !Permute
    to_bgr: false
    channel_first: true
  batch_transforms:
  - !PadBatch
    pad_to_stride: 32
    use_padded_im_info: false
  batch_size: 1
  shuffle: true
  worker_num: 2
  use_process: false

EvalReader:
  inputs_def:
    fields: ['image', 'im_info', 'im_id',  'im_shape', 'gt_bbox', 'gt_class', 'is_difficult', 'is_crowd']
  dataset:
    !VOCDataSet
    dataset_dir: /home/aistudio/work/CCTSDB_VOC1
    anno_path: /home/aistudio/work/CCTSDB_VOC1/test.txt
    use_default_label: false
  sample_transforms:
  - !DecodeImage
    to_rgb: true
    with_mixup: false
  - !NormalizeImage
    is_channel_first: false
    is_scale: true
    mean: [0.485,0.456,0.406]
    std: [0.229, 0.224,0.225]
  - !ResizeImage
    interp: 1
    max_size: 1333
    target_size: 800
    use_cv2: true
  - !Permute
    channel_first: true
    to_bgr: false
  batch_transforms:
  - !PadBatch
    pad_to_stride: 32
    use_padded_im_info: true
  batch_size: 1
  shuffle: false
  drop_empty: false
  worker_num: 2

TestReader:
  inputs_def:
    fields: ['image', 'im_info', 'im_id', 'im_shape']
  dataset:
    !ImageFolder
    anno_path: /home/aistudio/work/CCTSDB_VOC1/label_list.txt
    use_default_label: false
  sample_transforms:
  - !DecodeImage
    to_rgb: true
    with_mixup: false
  - !NormalizeImage
    is_channel_first: false
    is_scale: true
    mean: [0.485,0.456,0.406]
    std: [0.229, 0.224,0.225]
  - !ResizeImage
    interp: 1
    max_size: 1333
    target_size: 800
    use_cv2: true
  - !Permute
    channel_first: true
    to_bgr: false
  batch_transforms:
  - !PadBatch
    pad_to_stride: 32
    use_padded_im_info: true
  batch_size: 1
  shuffle: false

**Comment by willthefrog** [comment]
那后面这部分需要删掉吧
```yaml
TrainReader:
inputs_def:
fields: ['image', 'im_info', 'im_id', 'gt_bbox', 'gt_class','is_difficult']
dataset:
!VOCDataSet
dataset_dir: /home/aistudio/work/CCTSDB_VOC1/VOCdevkit/VOC2012/
anno_path: /home/aistudio/work/CCTSDB_VOC1/trainval.txt
use_default_label: false
#batch_size: 2

EvalReader:
inputs_def:
fields: ['image', 'im_info', 'im_id', 'im_shape', 'gt_bbox', 'gt_class', 'is_difficult', 'is_crowd']
dataset:
!VOCDataSet
dataset_dir: /home/aistudio/work/CCTSDB_VOC1/VOCdevkit/VOC2012/
anno_path: /home/aistudio/work/CCTSDB_VOC1/test.txt
use_default_label: false

TestReader:
dataset:
!ImageFolder
anno_path: /home/aistudio/work/CCTSDB_VOC1/label_list.txt
use_default_label: false
```

**Comment by Lxiaom123** [comment]
删掉了，还是同样的错误提示

**Comment by qingqing01** [comment]
@Lxiaom123 请问现在还是否有问题？ 从第一栏贴的配置看，是EvalReader的缩进有问题，EvalReader的缩进和TrainReader保持一致，而不是在TrainReader中。

**Comment by Lxiaom123** [comment]
@qingqing01 青青老师，上面的问题已解决，我还想问一下，VOC转COCO时，出现这样的问题如下
![image](https://user-images.githubusercontent.com/66930694/97830772-163deb80-1d09-11eb-86f5-c63f27595c61.png)

后面我改成不换行，出现的问题如下
![image](https://user-images.githubusercontent.com/66930694/97830738-ed1d5b00-1d08-11eb-9aca-0b4bb7ae47f9.png)
我确定这个文件是存在的，不知道咋弄了

**Comment by qingqing01** [comment]
@Lxiaom123 

看报错信息时， `unexpected indent`, 第6行的voc_label_list这里的缩进不是很正确。

第4、5行重复出现了voc_anno_list，一个一个转换吧，另外看下第5行后面是不是多了空格之类的。

**Comment by Lxiaom123** [comment]
@qingqing01 
青青老师，现在不报缩进错误了，新的错误如下：
![image](https://user-images.githubusercontent.com/66930694/97841834-8fe1d380-1d21-11eb-8fe2-9da873dab053.png)
但是这个文件我确定是存在的，给的是绝对路径，如下
![image](https://user-images.githubusercontent.com/66930694/97841796-80628a80-1d21-11eb-9288-7f09e2ee0c08.png)

**Comment by Lxiaom123** [comment]
@qingqing01
青青老师，现在不报缩进错误了，新的错误如下：
![image](https://user-images.githubusercontent.com/66930694/97842153-1eeeeb80-1d22-11eb-8e9f-dbccb4654e1f.png)

但是这个文件我确定是存在的，给的是绝对路径，如下
![image](https://user-images.githubusercontent.com/66930694/97842171-29a98080-1d22-11eb-8775-a0603241d0bb.png)

**Comment by Lxiaom123** [comment]
@qingqing01 按照这个方法试了，依旧还是提示这个错误

**Comment by yghstill** [comment]
@Lxiaom123 --voc_anno_list指定的只有文件名的列表。
比如：VOCdevkit/VOC2007/ImageSets/Main/trainval.txt
```
009881
009882
009884
009886
009887
...
```

**Comment by Lxiaom123** [comment]
@yghstill 按照您说的改了，又提示如下错误
![image](https://user-images.githubusercontent.com/66930694/97961428-47441c00-1dee-11eb-8041-1fe38467fa22.png)

**Comment by yghstill** [comment]
@Lxiaom123 请参考如下改法：https://github.com/PaddlePaddle/PaddleDetection/pull/1649/files#diff-adb42105e09aabcfe0cfa54440e20c1c680a4e22999a0d19612864bfb0ee11aeR236-R239

**Comment by yghstill** [comment]
fix in #1649
